src/CoreLoadEstimator.cc

```diff
@@ -0,0 +1,120 @@
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+#include "CoreLoadEstimator.h"
+
+namespace Arachne {
+
+CoreLoadEstimator::CoreLoadEstimator(int maxNumCores)
+    : lock(false),
+      utilizationThresholds(new double[maxNumCores]),
+      maxNumCores(maxNumCores) {}
+CoreLoadEstimator::~CoreLoadEstimator() { delete[] utilizationThresholds; }
+
+/**
+ * Returns -1,0,1 to suggest whether the core count should decrease,
+ * stay the same, or increase respectively.
+ */
+int
+CoreLoadEstimator::estimate(int curActiveCores) {
+    Lock guard(lock);
+    // Use collectionTime as a proxy to tell whether PerfStats have been
+    // previously recorded.
+    if (previousStats.collectionTime == 0) {
+        Arachne::PerfStats::collectStats(&previousStats);
+        return 0;
+    }
+    Arachne::PerfStats currentStats;
+    Arachne::PerfStats::collectStats(&currentStats);
+
+    // Evaluate idle time precentage multiplied by number of cores to
+    // determine whether we need to decrease the number of cores.
+    uint64_t idleCycles = currentStats.idleCycles - previousStats.idleCycles;
+    uint64_t totalCycles = currentStats.totalCycles - previousStats.totalCycles;
+    uint64_t utilizedCycles = totalCycles - idleCycles;
+    uint64_t totalMeasurementCycles = Cycles::fromNanoseconds(
+        currentStats.collectionTime - previousStats.collectionTime);
+    double totalUtilizedCores = static_cast<double>(utilizedCycles) /
+                                static_cast<double>(totalMeasurementCycles);
+
+    // Estimate load to determine whether we need to increment the number
+    // of cores.
+    uint64_t weightedLoadedCycles =
+        currentStats.weightedLoadedCycles - previousStats.weightedLoadedCycles;
+    double averageLoadFactor = static_cast<double>(weightedLoadedCycles) /
+                               static_cast<double>(totalCycles);
+    previousStats = currentStats;
+
+    if (estimationStrategy == LOAD_FACTOR) {
+        if (curActiveCores < maxNumCores &&
+            averageLoadFactor > loadFactorThreshold) {
+            // Record our current totalUtilizedCores, so we will only ramp down
+            // if utilization would drop below this level.
+            utilizationThresholds[curActiveCores] = totalUtilizedCores;
+            return 1;
+        }
+
+        // We should not ramp down if we have high occupancy of slots.
+        double averageNumSlotsUsed =
+            static_cast<double>(currentStats.numThreadsCreated -
+                                currentStats.numThreadsFinished) /
+            curActiveCores / Arachne::maxThreadsPerCore;
```

Comment:
The averageNumSlotsUsed name is confusing: this is the percentage of slots we've occupied. So it should not be a `Num` but an occupancy ratio.

Reply:
Done.